File: approaches/lifetime-objective/fast-engine/gate-trajectory.cpp
// Trajectory equivalence gate.
//
// Plays the same games twice, move for move, through the unmodified
// drop7::playHeadlessMove and through drop7::fast::playHeadlessMoveFast, and
// compares everything observable: board, next disc, score, score delta, the
// full wave list (depth, cleared, revealed, points), level advance, board
// clear, moves remaining, moves played and the terminal flag.
//
// Several policies are used, because a deterministic policy and a searching
// policy visit different parts of the state space:
//   --policy center   drop7::centerFirstMove
//   --policy d4       the frozen drop7::fair_only_depth4::chooseDepth4Action
//   --policy d3fast   depth-3 five-strata FastSearch, used only as a source of
//                     realistic long-game column sequences when the frozen
//                     depth-4 search is too expensive for the seed count.  The
//                     column is still chosen once and handed to both engines,
//                     so the gate remains a test of engine semantics only.
//
// The column played is always chosen from the reference state and handed to
// both engines, so the gate isolates engine semantics from policy semantics.
// Any nonzero mismatch count fails.

#include "slow-search.hpp"
#include "fast-engine.hpp"
#include "fast-search.hpp"
#include "corpus.hpp"

#include <atomic>
#include <iostream>
#include <mutex>
#include <string>
#include <thread>
#include <vector>

namespace {

using namespace drop7;
using namespace drop7::fast;

struct Report {
  std::uint64_t games = 0;
  std::uint64_t moves = 0;
  std::uint64_t waves = 0;
  std::uint64_t mismatches = 0;
  std::string first_failure;
};

bool sameWaves(const std::vector<Wave>& reference, const FullWaveSink& sink) {
  if (reference.size() != static_cast<std::size_t>(sink.count)) return false;
  for (std::size_t index = 0; index < reference.size(); ++index) {
    const Wave& left = reference[index];
    const Wave& right = sink.waves[index];
    if (left.depth != right.depth || left.cleared != right.cleared ||
        left.revealed != right.revealed || left.points != right.points) {
      return false;
    }
  }
  return true;
}

enum class Policy { kCenter, kDepth4, kDepth3Fast };

void compareGame(std::uint32_t seed, int maximum_moves, Policy policy,
                 FastSearch& helper, Report& report) {
  requireLease(seed);
  State reference = initialHeadlessState(seed);
  State fast = initialHeadlessState(seed);
  if (reference.board != fast.board || reference.next_disc != fast.next_disc) {
    ++report.mismatches;
    if (report.first_failure.empty()) {
      report.first_failure = "initial state differs";
    }
    return;
  }
  ++report.games;
  while (!reference.game_over && reference.moves_played < maximum_moves) {
    int column = 0;
    switch (policy) {
      case Policy::kCenter:
        column = centerFirstMove(reference.board);
        break;
      case Policy::kDepth4:
        column = ref::chooseDepth4Action(reference).action;
        break;
      case Policy::kDepth3Fast: {
        FastSearchMetrics metrics;
        column = helper.chooseAction(reference, metrics);
        break;
      }
    }
    if (column < 0 || !isLegal(reference.board, column)) {
      column = centerFirstMove(reference.board);
      if (column < 0) break;
    }
    MoveResult reference_move;
    FullWaveSink sink;
    FastMoveResult fast_move;
    const bool reference_ok =
        playHeadlessMove(reference, seed, column, reference_move);
    const bool fast_ok =
        playHeadlessMoveFast(fast, seed, column, sink, fast_move);
    ++report.moves;
    report.waves += reference_move.waves.size();

    const bool identical =
        reference_ok == fast_ok && reference.board == fast.board &&
        reference.next_disc == fast.next_disc &&
        reference.score == fast.score && reference.level == fast.level &&
        reference.moves_remaining == fast.moves_remaining &&
        reference.moves_played == fast.moves_played &&
        reference.game_over == fast.game_over &&
        reference_move.score_delta == fast_move.score_delta &&
        reference_move.cleared_board == fast_move.cleared_board &&
        reference_move.level_advanced == fast_move.level_advanced &&
        sameWaves(reference_move.waves, sink);
    if (!identical) {
      ++report.mismatches;
      if (report.first_failure.empty()) {
        report.first_failure = "seed 0x" + std::to_string(seed) + " move " +
                               std::to_string(reference.moves_played);
      }
      return;
    }
    if (!reference_ok) break;
  }
}

}  // namespace

int main(int argc, char** argv) {
  std::string policy = "center";
  int games = 4096;
  int maximum_moves = 2000;
  int threads = 8;
  std::uint32_t seed_start = 0;
  for (int index = 1; index + 1 < argc; index += 2) {
    const std::string key = argv[index];
    const std::string value = argv[index + 1];
    if (key == "--policy") policy = value;
    else if (key == "--games") games = std::stoi(value);
    else if (key == "--max-moves") maximum_moves = std::stoi(value);
    else if (key == "--threads") threads = std::stoi(value);
    else if (key == "--seed-start")
      seed_start = static_cast<std::uint32_t>(std::stoul(value, nullptr, 0));
  }
  Policy selected = Policy::kCenter;
  if (policy == "d4") selected = Policy::kDepth4;
  else if (policy == "d3fast") selected = Policy::kDepth3Fast;
  if (seed_start == 0) {
    seed_start = selected == Policy::kCenter ? kTrajectoryDeterministicSeeds
                                             : kTrajectorySearchSeeds;
  }
  requireLease(seed_start);
  requireLease(seed_start + static_cast<std::uint32_t>(games) - 1u);

  std::vector<Report> reports(static_cast<std::size_t>(threads));
  std::atomic<int> cursor{0};
  std::vector<std::thread> workers;
  for (int worker = 0; worker < threads; ++worker) {
    workers.emplace_back([&, worker] {
      Report& report = reports[static_cast<std::size_t>(worker)];
      FastSearchParameters helper_parameters;
      helper_parameters.depth = 3;
      helper_parameters.chance_samples = 5;
      helper_parameters.maximum_work = 3'200'000;
      helper_parameters.maximum_cache_entries = 60'000;
      FastSearch helper{helper_parameters};
      while (true) {
        const int game = cursor.fetch_add(1);
        if (game >= games) break;
        compareGame(seed_start + static_cast<std::uint32_t>(game),
                    maximum_moves, selected, helper, report);
      }
    });
  }
  for (auto& worker : workers) worker.join();

  Report total;
  for (const Report& report : reports) {
    total.games += report.games;
    total.moves += report.moves;
    total.waves += report.waves;
    total.mismatches += report.mismatches;
    if (total.first_failure.empty()) total.first_failure = report.first_failure;
  }
  std::cout << "policy: " << policy << "  seeds 0x" << std::hex << seed_start
            << "-0x" << (seed_start + static_cast<std::uint32_t>(games) - 1u)
            << std::dec << '\n';
  std::cout << "games compared: " << total.games << '\n';
  std::cout << "moves compared: " << total.moves << '\n';
  std::cout << "waves compared: " << total.waves << '\n';
  std::cout << "mismatches: " << total.mismatches << '\n';
  if (total.mismatches > 0) {
    std::cout << "first failure: " << total.first_failure << '\n';
  }
  const bool ok = total.mismatches == 0 && total.games > 0;
  std::cout << (ok ? "TRAJECTORY GATE PASSED\n" : "TRAJECTORY GATE FAILED\n");
  return ok ? 0 : 1;
}

File: approaches/fair-expectimax/rust-engine/src/bin/gate_trajectory.rs
// Trajectory equivalence gate: replays the columns recorded by the C++
// reference trace emitter (cpp/trace.cpp) through the Rust engine and
// compares everything observable, move by move: board, next disc, score,
// score delta, the full wave list, level, moves remaining, board-clear and
// level-advance flags, and the terminal flag.
//
// The columns come from the trace, so the gate isolates engine semantics
// from policy semantics.  Any nonzero mismatch count fails.
//
//   gate-trajectory --trace <path-to-c++-trace>

use drop7_rs::board::Board;
use drop7_rs::engine::{
    initial_ts_state, play_headless_move, play_ts_move, FullWaveSink, State,
};
use drop7_rs::rng::Mulberry32;

use std::env;
use std::fs;

struct MoveRecord {
    moves_played: i32,
    column: usize,
    score_delta: i64,
    board: String,
    next_disc: u8,
    score: i64,
    level: i32,
    moves_remaining: i32,
    game_over: bool,
    cleared: bool,
    advanced: bool,
    waves: String,
}

fn parse_move(line: &str) -> MoveRecord {
    // m <played> col <c> sd <delta> b <board> next <d> score <s> level <l>
    // mr <r> over <0|1> cleared <0|1> advanced <0|1> waves d:c:r:p ...
    let tokens: Vec<&str> = line.split_whitespace().collect();
    let mut record = MoveRecord {
        moves_played: 0,
        column: 0,
        score_delta: 0,
        board: String::new(),
        next_disc: 0,
        score: 0,
        level: 0,
        moves_remaining: 0,
        game_over: false,
        cleared: false,
        advanced: false,
        waves: String::new(),
    };
    let mut i = 0;
    while i < tokens.len() {
        match tokens[i] {
            "m" => {
                record.moves_played = tokens[i + 1].parse().unwrap();
                i += 2;
            }
            "col" => {
                record.column = tokens[i + 1].parse().unwrap();
                i += 2;
            }
            "sd" => {
                record.score_delta = tokens[i + 1].parse().unwrap();
                i += 2;
            }
            "b" => {
                record.board = tokens[i + 1].to_string();
                i += 2;
            }
            "next" => {
                record.next_disc = tokens[i + 1].parse().unwrap();
                i += 2;
            }
            "score" => {
                record.score = tokens[i + 1].parse().unwrap();
                i += 2;
            }
            "level" => {
                record.level = tokens[i + 1].parse().unwrap();
                i += 2;
            }
            "mr" => {
                record.moves_remaining = tokens[i + 1].parse().unwrap();
                i += 2;
            }
            "over" => {
                record.game_over = tokens[i + 1] == "1";
                i += 2;
            }
            "cleared" => {
                record.cleared = tokens[i + 1] == "1";
                i += 2;
            }
            "advanced" => {
                record.advanced = tokens[i + 1] == "1";
                i += 2;
            }
            "waves" => {
                record.waves = tokens[i + 1..].join(" ");
                break;
            }
            _ => panic!("unknown token {}", tokens[i]),
        }
    }
    record
}

fn main() {
    let mut trace_path = String::new();
    let mut driver = String::from("headless");
    let args: Vec<String> = env::args().collect();
    let mut i = 1;
    while i + 1 < args.len() {
        match args[i].as_str() {
            "--trace" => trace_path = args[i + 1].clone(),
            "--driver" => driver = args[i + 1].clone(),
            _ => {}
        }
        i += 2;
    }
    if trace_path.is_empty() {
        eprintln!("usage: gate-trajectory --trace <trace file> [--driver headless|ts]");
        std::process::exit(2);
    }
    let text = fs::read_to_string(&trace_path).expect("read trace");

    let mut games = 0u64;
    let mut moves = 0u64;
    let mut waves_total = 0u64;
    let mut mismatches = 0u64;
    let mut first_failure = String::new();

    let mut state = State {
        board: Board::initial(),
        next_disc: 0,
        score: 0,
        level: 1,
        moves_remaining: 5,
        moves_played: 0,
        game_over: false,
    };
    let mut seed = 0u32;
    let mut sink = FullWaveSink::new();
    let mut ts_random = Mulberry32::new(0);

    for line in text.lines() {
        if let Some(rest) = line.strip_prefix("game 0x") {
            // game <seed> next <d>
            let tokens: Vec<&str> = rest.split_whitespace().collect();
            seed = u32::from_str_radix(tokens[0], 16).unwrap();
            let next: u8 = tokens[2].parse().unwrap();
            if driver == "ts" {
                ts_random = Mulberry32::new(seed);
                state = initial_ts_state(&mut ts_random);
            } else {
                state = State::initial_headless(seed);
            }
            if state.next_disc != next {
                mismatches += 1;
                if first_failure.is_empty() {
                    first_failure = format!("seed {seed:#x}: initial next disc differs");
                }
            }
            games += 1;
        } else if line.starts_with("m ") {
            let record = parse_move(line);
            sink.clear();
            let result = if driver == "ts" {
                play_ts_move(&mut state, &mut ts_random, record.column, &mut sink)
            } else {
                play_headless_move(&mut state, seed, record.column, &mut sink)
            };
            moves += 1;
            waves_total += sink.count as u64;
            let rust_waves: Vec<String> = sink.waves[..sink.count]
                .iter()
                .map(|w| format!("{}:{}:{}:{}", w.depth, w.cleared, w.revealed, w.points))
                .collect();
            let identical = result.is_some()
                && state.moves_played == record.moves_played
                && result.unwrap().score_delta == record.score_delta
                && state.board.serialize() == record.board
                && state.next_disc == record.next_disc
                && state.score == record.score
                && state.level == record.level
                && state.moves_remaining == record.moves_remaining
                && state.game_over == record.game_over
                && result.unwrap().cleared_board == record.cleared
                && result.unwrap().level_advanced == record.advanced
                && rust_waves.join(" ") == record.waves;
            if !identical {
                mismatches += 1;
                if first_failure.is_empty() {
                    first_failure = format!(
                        "seed {seed:#x} move {}\n  rust: b {} next {} score {} mr {} waves {}\n  ref:  b {} next {} score {} mr {} waves {}",
                        record.moves_played,
                        state.board.serialize(),
                        state.next_disc,
                        state.score,
                        state.moves_remaining,
                        rust_waves.join(" "),
                        record.board,
                        record.next_disc,
                        record.score,
                        record.moves_remaining,
                        record.waves,
                    );
                }
            }
        } else if line.starts_with("end ") {
            // end score <s> moves <m> over <0|1>
            let tokens: Vec<&str> = line.split_whitespace().collect();
            let end_score: i64 = tokens[2].parse().unwrap();
            let end_moves: i32 = tokens[4].parse().unwrap();
            let end_over = tokens[6] == "1";
            if state.score != end_score
                || state.moves_played != end_moves
                || state.game_over != end_over
            {
                mismatches += 1;
                if first_failure.is_empty() {
                    first_failure = format!("seed {seed:#x}: terminal state differs");
                }
            }
        }
    }

    println!("games compared: {games}");
    println!("moves compared: {moves}");
    println!("waves compared: {waves_total}");
    println!("mismatches: {mismatches}");
    if mismatches > 0 {
        println!("first failure: {first_failure}");
    }
    let ok = mismatches == 0 && games > 0;
    println!(
        "{}",
        if ok {
            "TRAJECTORY GATE PASSED"
        } else {
            "TRAJECTORY GATE FAILED"
        }
    );
    std::process::exit(if ok { 0 } else { 1 });
}
